### PR TITLE
Feature/tec 3630 smooth transition v1 to v2

### PR DESCRIPTION
**Pull request by jesseeproductions**
_Ref:_ [TEC-3630](https://moderntribe.atlassian.net/browse/TEC-3630)
_Ref:_ [TEC-3593](https://moderntribe.atlassian.net/browse/TEC-3593)
_Ref:_ [ECP-517](https://moderntribe.atlassian.net/browse/ECP-517)

Related: https://github.com/moderntribe/events-pro/pull/1607

📸 
**TEC-only**
_master_: http://p.tri.be/Z6CPP4 & http://p.tri.be/AZbESz
_branch_: http://p.tri.be/bubE2T & http://p.tri.be/iEqQTn

**TEC & ECP**
master: http://p.tri.be/ADkHbn && http://p.tri.be/7wzpTo

ech...here's a video: http://p.tri.be/IsP4Bm

### CODE REVIEW

File: src/Tribe/Views/V2/Widgets/Compatibility.php [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+	/**
+	 * Switches the primary and alternative id base when v1 is active
+	 * or v2 widgets are disabled, this enables support for upgrading from
+	 * v1 free list widget to the v1 Pro advanced list widget.
+	 *
+	 * @since TBD
+	 *
+	 */
+	public function switch_to_v1_compatibility() {
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/Widgets/Compatibility.php:55]: :no_entry_sign: Additional blank lines found at end of doc comment

<sub>hash: be297244ace0dc2fd73ffa29408cbace</sub>

File: src/Tribe/Views/V2/Widgets/Compatibility.php [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+		// if Pro is disabled, use the defaults.
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/Widgets/Compatibility.php:56]: * :no_entry_sign: Please, make sure that a callback to `'tribe_plugins_loaded'` filter is returning void intentionally.
* :no_entry_sign: Please, make sure that a callback to `'tribe_plugins_loaded'` filter is always returning some value.

<sub>hash: c27f54fa50c2ebb540197cdf751691e0</sub>

File: src/Tribe/Views/V2/Widgets/Service_Provider.php [hunk 3]
```diff
@@ -37,18 +37,32 @@ class Service_Provider extends \tad_DI52_ServiceProvider {
...
 		$this->hook();
 	}
 
+	/**
+	 * Registers the provider handling for compatibility hooks.
+	 *
+	 * @since TBD
+	 */
+	protected function register_compatibility() {
+		$this->container->singleton( Compatibility::class, Compatibility::class, ['hooks'] );
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/Widgets/Service_Provider.php:61]: * :no_entry_sign: Missing space after array opener.
* :no_entry_sign: Missing space after array opener.
* :no_entry_sign: Missing space before array closer.
* :no_entry_sign: Missing space before array closer.

<sub>hash: ecc87719c66e9379c3c03b4d70ee2fcf</sub>

File: src/Tribe/Views/V2/Widgets/Service_Provider.php [hunk 4]
```diff
@@ -37,18 +37,32 @@ class Service_Provider extends \tad_DI52_ServiceProvider {
...
+		$this->container->singleton( 'pro.views.v2.widgets.compatibility', Compatibility::class, ['hooks'] );
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/Widgets/Service_Provider.php:62]: * :no_entry_sign: Missing space after array opener.
* :no_entry_sign: Missing space after array opener.
* :no_entry_sign: Missing space before array closer.
* :no_entry_sign: Missing space before array closer.

<sub>hash: 9e205c5a88e9d6acaa5a2b9ed7fcf57c</sub>

**Review by tr1b0t — COMMENTED** [review]
:no_entry_sign: 11 error(s)

File: src/Tribe/Views/V2/Widgets/Compatibility.php [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+	/**
+	 * Switches the primary and alternative id base when v1 is active
+	 * or v2 widgets are disabled, this enables support for upgrading from
+	 * v1 free list widget to the v1 Pro advanced list widget.
+	 *
+	 * @since TBD
+	 */
+	public function switch_compatibility() {
+		// if Pro is disabled, use the defaults.
```
Comment (lucatume) [inline src/Tribe/Views/V2/Widgets/Compatibility.php:60]:
```suggestion
		// If Pro is disabled, use the defaults.
```

File: src/Tribe/Views/V2/Widgets/Compatibility.php [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+	 * Remap the widget id_base for the Pro Advanced List Widget.
+	 *
+	 * @since TBD
+	 *
+	 * @param array<string,mixed> $widget_areas An array of widgets areas with the saved widgets in each location.
+	 *
+	 * @return array<string,mixed> $widget_areas An array of widgets areas with the saved widgets in each location.
+	 */
+	public function remap_list_widget_id_bases( $widget_areas ) {
+
```
Comment (lucatume) [inline src/Tribe/Views/V2/Widgets/Compatibility.php:86]:
```suggestion
```

Here and elsewhere, remove blank lines at start and end of function and methods definitions.

File: src/Tribe/Views/V2/Widgets/Compatibility.php [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+		$widget_areas_bak = $widget_areas;
+
+		if ( ! is_array( $widget_areas ) ) {
+			return $widget_areas;
+		}
+
+		foreach ( $widget_areas as $key => $widget_location ) {
+
```
Comment (lucatume) [inline src/Tribe/Views/V2/Widgets/Compatibility.php:94]: Same removal of blank lines of blocks, here and elsewhere.

File: src/Tribe/Views/V2/Widgets/Compatibility.php [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+	 */
+	public function hooks() {
+		add_action( 'tribe_plugins_loaded', [ $this, 'switch_compatibility' ] );
+		add_filter( 'option_sidebars_widgets', [ $this, 'remap_list_widget_id_bases' ] );
+	}
+
+	/**
+	 * Switches the primary and alternative id base when v1 is active
+	 * or v2 widgets are disabled, this enables support for upgrading from
+	 * v1 free list widget to the v1 Pro advanced list widget.
```
Comment (lucatume) [inline src/Tribe/Views/V2/Widgets/Compatibility.php:55]: Please, add a note here about why this method should live in TEC and not PRO, and why.
It's not an immediate intuition and someone with good intentions might think to move it in the future.

**Review by lucatume — CHANGES_REQUESTED** [review]
Please, add tests to cover the option filtering: having a clear indication of what is happening and what the filtering does might make this on-the-fly migration easier to understand and cover this piece of code from regression.

File: src/Tribe/Views/V2/Widgets/Compatibility.php [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+			return $widgets;
+		}
+
+		// Get the saved alternative widgets.
+		$alternative_options = get_option( "widget_{$this->alternative_id_base}" );
+		if ( ! is_array( $alternative_options ) ) {
+			return $widgets;
+		}
+
+		// Combine arrays and key the array keys.
```
Comment (Camwyn) [inline src/Tribe/Views/V2/Widgets/Compatibility.php:128]:
```suggestion
		// Combine arrays and keep the array keys.
```
?

File: src/Tribe/Views/V2/Widgets/Compatibility.php [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+	}
+
+	/**
+	 * Merge the Event List and Advanced List Widget Options.
+	 *
+	 * @since TBD
+	 *
+	 * @param array<int,mixed> $widgets An array of saved widgets.
+	 *
+	 * @return array<int,mixed> $widgets An array of saved widgets.
```
Comment (Camwyn) [inline src/Tribe/Views/V2/Widgets/Compatibility.php:114]:
```suggestion
	 * @return array<int,mixed> $widgets The modified array of saved widgets.
```

**Review by Camwyn — CHANGES_REQUESTED** [review]

**Review by lucatume — APPROVED** [review]

**Review by Camwyn — DISMISSED** [review]

File: src/admin-views/widgets/components/radio.php [hunk 11]
```diff
@@ -29,15 +29,15 @@ class="tribe-widget-form-control tribe-widget-form-control--radio"
 >
 	<input
 		class="tribe-widget-form-control__input"
-		id="<?php echo esc_attr( $id ); ?>"
+		id="<?php echo esc_attr( $id ) . '-' . strtolower( esc_attr( $button_value ) ); ?>"
```
Comment (tr1b0t) [inline src/admin-views/widgets/components/radio.php:32]: * :no_entry_sign: All output should be run through an escaping function (see the Security sections in the WordPress Developer Handbooks), found 'strtolower'.
* :warning: Expected next thing to be an escaping function (see Codex for 'Data Validation'), not ')'

<sub>hash: d868937ea6dceb2e75d0e1c7f682efc1</sub>

File: src/admin-views/widgets/components/radio.php [hunk 12]
```diff
@@ -29,15 +29,15 @@ class="tribe-widget-form-control tribe-widget-form-control--radio"
...
 		name="<?php echo esc_attr( $name ); ?>"
 		type="radio"
 		value="<?php echo esc_attr( $button_value ); ?>"
 		<?php checked( $button_value, $value ); ?>
 	/>
 	<label
 		class="tribe-widget-form-control__label"
-		for="<?php echo esc_attr( $id ); ?>"
+		for="<?php echo esc_attr( $id ) . '-' . strtolower( esc_attr( $button_value ) ); ?>"
```
Comment (tr1b0t) [inline src/admin-views/widgets/components/radio.php:40]: * :no_entry_sign: All output should be run through an escaping function (see the Security sections in the WordPress Developer Handbooks), found 'strtolower'.
* :warning: Expected next thing to be an escaping function (see Codex for 'Data Validation'), not ')'

<sub>hash: 9f79e689d085f79ffebc163a0c94cf4c</sub>

**Review by tr1b0t — COMMENTED** [review]
:no_entry_sign: 2 error(s)
:warning: 2 warning(s)

**Review by Camwyn — APPROVED** [review]

**Pull request merged**